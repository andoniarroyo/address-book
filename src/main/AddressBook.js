// @flow
// external modules
import React from 'react';
import { Link } from 'react-router-dom';
// internal types
import type { ContactListType, ContactType } from './index';
import type { RemoveContactActionFactoryType } from './actions';
// internal modules
import ContactRow from './components/ContactRow';
import EmptyContact from './components/EmptyContacts';
// styles
import './styles/addressBook.scss';

export type AddressBookPropsType = {
  contactList: ContactListType,
  removeContact: RemoveContactActionFactoryType,
};

/**
 * The contact rows render method
 * @param {ContactListType} contactList The list of contacts to be rendered
 * @param {RemoveContactActionFactoryType} removeContact The remove contact action factory
 * @returns {React$Element<any>} HTML markup for the list of contacts
 * @private
 */
const renderContacts = (contactList: ContactListType, removeContact: RemoveContactActionFactoryType)
: Array<React.Element<*>> => (
  contactList.map((contact: ContactType): React.Element<*> => (
    <ContactRow key={contact.id} contact={contact} removeContact={removeContact} />
  ))
);

/**
 * The adress book render method
 * @param {AddressBookPropsType} props The properties to render the object
 * @returns {React$Element<any>} HTML markup for the component
 * @public
 */
const AddressBook = ({ contactList, removeContact }: AddressBookPropsType): React.Element<*> => {
  const contacts = (!contactList.length) ? <EmptyContact /> : renderContacts(contactList, removeContact);
  return (
    <main>
      <header>
        <h3>Address book</h3>
        <Link to="/add">
            Create a new contact
        </Link>
      </header>
      <section className="contact-list">
        {contacts}
      </section>
    </main>
  );
};

export default AddressBook;
