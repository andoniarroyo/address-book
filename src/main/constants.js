// @flow
// paths
export const HOME_PATH = '/';
export const INVALID_CONTACT_PATH = '/invalid';

// contact info modes
export const ADD_MODE = 'ADD';
export const EDIT_MODE = 'EDIT';
// actions related constants
export const CONTACT_ADDED = 'CONTACT_ADDED';
export const CONTACT_REMOVED = 'CONTACT_REMOVED';
export const CONTACT_UPDATED = 'CONTACT_UPDATED';
// settings
export const DEFAULT_COUNTRY_CODE = 'GB';
