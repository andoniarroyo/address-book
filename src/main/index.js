// @flow
// internal modules
import contactListReducer from './reducers';

export type ContactType = {
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    countryCode: string,
}

export type ContactPartialType = {
  id?: string,
  firstName?: string,
  lastName?: string,
  email?: string,
  countryCode?: string,
}

export type ContactListType = Array<ContactType>;

export type StateType = {
  contacts: ContactListType
}

const reducers = {
  contacts: contactListReducer,
};

export default reducers;
