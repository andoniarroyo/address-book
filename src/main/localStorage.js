// @flow
import type { StateType } from './index';

/**
 * Loads the state from the local storage
 * @returns {?StateType} The saved state
 * @public
 */
export const loadState = (): ?StateType => {
  const state = localStorage.getItem('state');
  return state ? JSON.parse(state) : null;
};

/**
 * Saves the state to the local storage
 * @param {StateType} state The state to be saved
 * @returns {void}
 * @public
 */
export const saveState = (state: StateType): void => localStorage.setItem('state', JSON.stringify(state));
