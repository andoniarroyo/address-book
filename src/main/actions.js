// @flow
// internal types
import type { ContactType } from './index';
// internal modules
import { CONTACT_ADDED, CONTACT_REMOVED, CONTACT_UPDATED } from './constants';

export type CommonActionType = {
  type: string,
}

export type AddContactActionType = {
  type: string,
  newContact: ContactType
};

export type AddContactActionFactoryType = (newContact: ContactType) => AddContactActionType;

export type RemoveContactActionType = {
  type: string,
  contactID: string
};

export type RemoveContactActionFactoryType = (contactID: string) => RemoveContactActionType;

export type UpdateContactActionType = {
  type: string,
  updatedContact: ContactType
};

export type UpdateContactActionFactoryType = (updatedContact: ContactType) => UpdateContactActionType;

/**
 * Action creator to add a new contact
 * @param {ContactType} newContact The new contact to be added
 * @returns {AddContactActionType} The generated action
 * @public
 */
export const addContact = (newContact: ContactType): AddContactActionType => (
  {
    type: CONTACT_ADDED,
    newContact,
  }
);

/**
 * Action creator to remove a new contact
 * @param {string} contactID The identifier of the contact to be removed
 * @returns {RemoveContactActionType} The generated action
 * @public
 */
export const removeContact = (contactID: string) : RemoveContactActionType => (
  {
    type: CONTACT_REMOVED,
    contactID,
  }
);

/**
 * Action creator to update a contact
 * @param {ContactType} updatedContact The contact to be updated with the new information
 * @returns {UpdateContactActionType} The generated action
 * @public
 */
export const updateContact = (updatedContact: ContactType) : UpdateContactActionType => (
  {
    type: CONTACT_UPDATED,
    updatedContact,
  }
);
