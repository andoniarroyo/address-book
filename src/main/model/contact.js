// @flow
// external modules
import shortid from 'shortid';
// internal types
import type { ContactPartialType, ContactType } from '../index';
// internal modules
import { DEFAULT_COUNTRY_CODE } from '../constants';

/**
 * Checks if the incoming string exists and is longer than 0
 * @param {string} stringToValidate The string to be evaluated
 * @return {boolean} True if the string is longer than 0, False otherwise
 * @private
 */
const existingString = (stringToValidate: string): boolean => !!(stringToValidate);

/**
 * Checks if the email reported is a valid email
 * @param {string} emailToValidate Email account to be evaluated
 * @return {boolean} True if the email is valid, False otherwise
 * @private
 */
const isValidEmail = (emailToValidate: string): boolean => (
  new RegExp(/(\w+)@(\w+).[a-zA-Z]/g).test(emailToValidate)
);

/**
 * Checks if the incoming contact has a valid values for the required properties
 * @param {ContactType} contact The contact to be evaluated
 * @return {boolean} True if the contact is valid, False otherwise
 * @public
 */
export const isValidContact = (contact: ContactType): boolean => (
  Boolean(contact && existingString(contact.id) && existingString(contact.firstName)
    && existingString(contact.lastName) && isValidEmail(contact.email) && existingString(contact.countryCode))
);

/**
 * Creates a new contact with the default values for the non informed properties
 * @param {ContactPartialType} contactPartial Contact partially informed to be used as refeence
*  @return {ContactType} The created contact
 * @public
 */
export const createContact = ({
  id = shortid.generate(), firstName = '', lastName = '', email = '', countryCode = DEFAULT_COUNTRY_CODE,
}
: ContactPartialType): ContactType => (
  {
    id, firstName, lastName, email, countryCode,
  }
);
