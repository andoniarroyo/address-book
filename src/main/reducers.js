// @flow
// internal types
import type {
  CommonActionType,
  AddContactActionType,
  RemoveContactActionType,
  UpdateContactActionType,
} from './actions';
import type { ContactListType, ContactType } from './index';
// internal modules
import { CONTACT_ADDED, CONTACT_REMOVED, CONTACT_UPDATED } from './constants';

/**
 * Sorts the incoming contact list based on the last name values
 * @param {ContactListType} contactList The list to be sorted
 * @returns {ContactListType} The ordered list
 * @private
 */
const sortContactList = (contactList: ContactListType): ContactListType => (
  contactList.sort((contact1: ContactType, contact2:ContactType): number => (
    contact1.lastName.toLowerCase().localeCompare(contact2.lastName.toLowerCase())
  ))
);

/**
 * Adds a contact to the end of the contact list
 * @param {ContactListType} contactList The contact list to add the contact
 * @param {ContactType} newContact The new contact to be added
 * @returns {ContactListType} The new generated contact list, including the new contact
 * @private
 */
const addContact = (contactList: ContactListType, newContact: ContactType): ContactListType => (
  contactList.concat([newContact])
);

/**
 * Removes the contact from the contact list (if it exists)
 * @param {ContactListType} contactList The contact list to be removed the contact from
 * @param {string} contactID The identifier of the contact to be removed
 * @returns {ContactListType} The new generated contact list, removing the contact
 * @private
 */
const removeContact = (contactList: ContactListType, contactID: string): ContactListType => (
  contactList.filter((contact: ContactType): boolean => contact.id !== contactID)
);

/**
 * Updates the contact from the contact list (if it exists)
 * @param {ContactListType} contactList The contact list to be removed the contact from
 * @param {ContactType} updatedContact The modified contact with the updated information
 * @returns {ContactListType} The new generated contact list, updating the contact
 * @private
 */
const updateContact = (contactList: ContactListType, updatedContact: ContactType): ContactListType => (
  contactList.map((contact: ContactType): ContactType => (contact.id === updatedContact.id ? updatedContact : contact))
);

/**
 * Reduces the contact list adding a new contact
 * @param {ContactListType} contactList The current contact list
 * @param {AddContactActionType} action The add contact action
 * @returns {ContactListType} The new contact list including the new contact
 * @private
 */
const contactAddedReducer = (contactList: ContactListType = [], action: AddContactActionType)
: ContactListType => (
  action.type === CONTACT_ADDED ? sortContactList(addContact(contactList, action.newContact)) : contactList
);

/**
 * Reduces the contact list removing the contact in case it exists
 * @param {ContactListType} contactList The current contact list
 * @param {RemoveContactActionType} action The remove contact action
 * @returns {ContactListType} The new contact removing the specified contact
 * @private
 */
const contactRemovedReducer = (contactList: ContactListType = [], action: RemoveContactActionType)
: ContactListType => (
  action.type === CONTACT_REMOVED ? removeContact(contactList, action.contactID) : contactList
);

/**
 * Reduces the contact list updating the contact in case it exists
 * @param {ContactListType} contactList The current contact list
 * @param {UpdateContactActionType} action The update contact action
 * @returns {ContactListType} The new contact list modifying the contact to be updated
 * @private
 */
const contactUpdatedReducer = (contactList: ContactListType = [], action: UpdateContactActionType)
: ContactListType => (
  action.type === CONTACT_UPDATED ? updateContact(contactList, action.updatedContact) : contactList
);

const actionHandlers = {
  CONTACT_ADDED: contactAddedReducer,
  CONTACT_REMOVED: contactRemovedReducer,
  CONTACT_UPDATED: contactUpdatedReducer,
};

/**
 * Reduces the contact list based on the action
 * @param {ContactListType} contactList The current contact list
 * @param {CommonActionType} action The action to be handled
 * @returns {ContactListType} The new contact list modifying based on the incoming action
 * @public
 */
const contactListReducer = (contactList: ContactListType = [], action: CommonActionType) => {
  const handler = actionHandlers[action.type];
  if (handler) {
    return handler(contactList, action);
  }
  return contactList;
};

export default contactListReducer;
