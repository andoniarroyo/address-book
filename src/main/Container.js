// @flow
// external modules
import { connect } from 'react-redux';
// internal types
import type { StateType, ContactListType } from './index';
// internal modules
import { contactListSelector } from './selectors';
import { addContact, removeContact, updateContact } from './actions';
import AddressBook from './AddressBook';

type StateToPropsType = {
  contactList: ContactListType
}

/**
 * Maps the application state to props passed to the component
 * @param {StateType} state The application redux state
 * @returns {StateToPropsType} The props coming from the state
 * @private
 */
const mapStateToProps = (state: StateType): StateToPropsType => ({
  contactList: contactListSelector(state),
});

const mapDispatchToProps = {
  addContact,
  removeContact,
  updateContact,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddressBook);
