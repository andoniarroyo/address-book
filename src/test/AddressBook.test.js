// external modules
import React from 'react';
import renderer from 'react-test-renderer';
// internal modules
import { createContact } from '../main/model/contact';
import { removeContact } from '../main/actions';
import AddressBook from '../main/AddressBook';
import staticRouterWrapper from './utils/staticRouterWrapper';

describe('<AddressBook />', () => {
  it('renders the component', () => {
    const props = {
      contactList: [
        createContact({
          id: '1', firstName: 'Andoni', lastName: 'Arroyo', email: 'andoni.arroyo@gmail.com', countryCode: 'GB',
        }),
        createContact({
          id: '2', firstName: 'Naiara', lastName: 'Saratxaga', email: 'nsaratx@gmail.com', countryCode: 'ES',
        }),
      ],
      removeContact,
    };

    const tree = renderer.create(staticRouterWrapper(<AddressBook {...props} />)).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
