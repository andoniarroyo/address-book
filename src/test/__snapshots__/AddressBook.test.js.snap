// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressBook /> renders the component 1`] = `
<main>
  <header>
    <h3>
      Address book
    </h3>
    <a
      href="/add"
      onClick={[Function]}
    >
      Create a new contact
    </a>
  </header>
  <section
    className="contact-list"
  >
    <article
      className="contact"
    >
      <span
        className="item"
      >
        <a
          href="/edit/1"
          onClick={[Function]}
        >
          Andoni
        </a>
      </span>
      <span
        className="item"
      >
        Arroyo
      </span>
      <span
        className="item"
      >
        <a
          href="mailto:andoni.arroyo@gmail.com"
        >
          andoni.arroyo@gmail.com
        </a>
      </span>
      <span
        className="item"
      >
        United Kingdom
      </span>
      <span
        className="item item-final"
      >
        <button
          id="remove-button"
          onClick={[Function]}
        >
          Remove it!
        </button>
      </span>
    </article>
    <article
      className="contact"
    >
      <span
        className="item"
      >
        <a
          href="/edit/2"
          onClick={[Function]}
        >
          Naiara
        </a>
      </span>
      <span
        className="item"
      >
        Saratxaga
      </span>
      <span
        className="item"
      >
        <a
          href="mailto:nsaratx@gmail.com"
        >
          nsaratx@gmail.com
        </a>
      </span>
      <span
        className="item"
      >
        Spain
      </span>
      <span
        className="item item-final"
      >
        <button
          id="remove-button"
          onClick={[Function]}
        >
          Remove it!
        </button>
      </span>
    </article>
  </section>
</main>
`;
